File: tradesmanager/android/app/src/main/java/il/co/tradesmanager/data/catalog/CatalogSource.kt
package il.co.tradesmanager.data.catalog

import android.content.Context
import java.io.File
import kotlinx.serialization.json.Json

/**
 * Where catalogue files are read from.
 *
 * The app ships a complete copy in its assets, so a phone that has never had a
 * signal still onboards with full catalogues. An optional cloud refresh writes
 * a newer set into the app's own files directory; this class prefers that copy
 * when its catalogVersion is higher, which is how catalogues can be corrected
 * between store releases without an app update.
 */
class CatalogSource(
    private val context: Context,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {

    private val downloadedRoot: File
        get() = File(context.filesDir, DOWNLOADED_DIR)

    fun manifest(): CatalogManifest {
        val bundled = json.decodeFromString<CatalogManifest>(readAsset(MANIFEST))
        val downloaded = runCatching {
            json.decodeFromString<CatalogManifest>(File(downloadedRoot, MANIFEST).readText())
        }.getOrNull()
        return when {
            downloaded == null -> bundled
            downloaded.schemaVersion != bundled.schemaVersion -> bundled
            downloaded.catalogVersion > bundled.catalogVersion -> downloaded
            else -> bundled
        }
    }

    inline fun <reified T> decode(relativePath: String): T =
        jsonParser.decodeFromString(readText(relativePath))

    /** Downloaded copy wins per file, so a partial refresh is still coherent. */
    fun readText(relativePath: String): String {
        val downloaded = File(downloadedRoot, "$ROOT/$relativePath")
        return if (downloaded.isFile) downloaded.readText() else readAsset("$ROOT/$relativePath")
    }

    private fun readAsset(path: String): String =
        context.assets.open(path).bufferedReader().use { it.readText() }

    @PublishedApi
    internal val jsonParser: Json get() = json

    companion object {
        const val ROOT = "catalog"
        const val MANIFEST = "catalog/manifest.json"
        const val DOWNLOADED_DIR = "catalog-update"
    }
}

File: tradesmanager/ios/TradesManager/Shared/CatalogSource.swift
import Foundation

/// Reads the catalogues that ship inside the app bundle, preferring a newer
/// set downloaded into Application Support.
///
/// This is the same rule the Android `CatalogSource` follows, and for the same
/// reason: a phone that has never had a signal must still onboard with full
/// catalogues, while a correction between store releases must be able to reach
/// devices without an app update.
public struct CatalogSource {

    public enum LoadError: Error {
        case missingFile(String)
        case unsupportedSchema(found: Int, supported: Int)
    }

    /// The schema this build understands. A downloaded catalogue declaring a
    /// different one is ignored rather than half-read.
    public static let supportedSchemaVersion = 1

    private static let root = "catalog"

    private let bundle: Bundle
    private let downloadedRoot: URL?
    private let decoder = JSONDecoder()

    public init(bundle: Bundle = .main, downloadedRoot: URL? = CatalogSource.defaultDownloadRoot()) {
        self.bundle = bundle
        self.downloadedRoot = downloadedRoot
    }

    public static func defaultDownloadRoot() -> URL? {
        try? FileManager.default
            .url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
            .appendingPathComponent("catalog-update", isDirectory: true)
    }

    public func manifest() throws -> CatalogManifest {
        let bundled: CatalogManifest = try decode(relativePath: "manifest.json", allowDownloaded: false)
        guard bundled.schemaVersion == Self.supportedSchemaVersion else {
            throw LoadError.unsupportedSchema(found: bundled.schemaVersion, supported: Self.supportedSchemaVersion)
        }

        guard let downloaded: CatalogManifest = try? decode(relativePath: "manifest.json", allowDownloaded: true),
              downloaded.schemaVersion == Self.supportedSchemaVersion,
              downloaded.catalogVersion > bundled.catalogVersion
        else {
            return bundled
        }
        return downloaded
    }

    public func items(for trade: CatalogTrade) throws -> CatalogItemFile {
        try decode(relativePath: trade.itemsFile)
    }

    public func safety(for trade: CatalogTrade) throws -> SafetyFile? {
        guard let path = trade.safetyFile else { return nil }
        return try decode(relativePath: path)
    }

    public func templates(for trade: CatalogTrade) throws -> TemplateFile? {
        guard let path = trade.templatesFile else { return nil }
        return try decode(relativePath: path)
    }

    private func decode<T: Decodable>(relativePath: String, allowDownloaded: Bool = true) throws -> T {
        try decoder.decode(T.self, from: try data(relativePath: relativePath, allowDownloaded: allowDownloaded))
    }

    /// Downloaded copy wins per file, so a partial refresh is still coherent.
    private func data(relativePath: String, allowDownloaded: Bool) throws -> Data {
        if allowDownloaded,
           let downloadedRoot,
           case let candidate = downloadedRoot.appendingPathComponent("\(Self.root)/\(relativePath)"),
           FileManager.default.fileExists(atPath: candidate.path) {
            return try Data(contentsOf: candidate)
        }
        guard let url = bundle.url(forResource: "\(Self.root)/\(relativePath)", withExtension: nil) else {
            throw LoadError.missingFile(relativePath)
        }
        return try Data(contentsOf: url)
    }
}
